docker/Dockerfile

```diff
@@ -1,101 +1,125 @@
- && git submodule update \
- # verilator model
- && cd /src/target/snitch_cluster \
- && make CFG_OVERRIDE=cfg/fdiv.hjson bin/snitch_cluster.vlt \
- # spike-dasm
- && cd /src/target/snitch_cluster/work-vlt/riscv-isa-sim \
- && ./configure --prefix=/opt/snitch-spike \
- && make install \
- # snitch runtime
- && cd /src/target/snitch_cluster \
- && make DEBUG=ON sw \
- # clang+llvm+lld
- && mkdir -p /opt/snitch-llvm \
- && wget -qO- https://github.com/pulp-platform/llvm-project/releases/download/0.12.0/riscv32-pulp-llvm-ubuntu1804-0.12.0.tar.gz | \
-    tar xz --strip-components=1 -C /opt/snitch-llvm \
- # python 3.11
- && apt update -y \
- && apt install -y \
-      build-essential \
-      zlib1g-dev \
-      libncurses5-dev \
-      libgdbm-dev \
-      libnss3-dev \
-      libssl-dev \
-      libreadline-dev \
-      libffi-dev \
-      libsqlite3-dev \
-      wget \
-      libbz2-dev \
- && mkdir -p /tmp/python-src \
- && wget -qO- https://www.python.org/ftp/python/3.11.5/Python-3.11.5.tgz | \
-    tar xz --strip-components=1 -C /tmp/python-src \
- && cd /tmp/python-src \
- && ./configure --prefix=/opt/python3.11 \
- && make install
-
-FROM ubuntu:18.04 as toolchain
+#-------------------------------------------------------------------------------
+# Stage 1: Builder
+# Clones the repository and builds the necessary artifacts.
+#-------------------------------------------------------------------------------
+FROM ghcr.io/pulp-platform/snitch_cluster:main AS builder
+
+# Clone the snitch_cluster repository into /repo
+RUN git clone -b fix-deps https://github.com/pulp-platform/snitch_cluster /repo
+
+WORKDIR /repo
+
+# Patch .gitmodules to replace SSH URLs with HTTPS
+RUN sed -i 's|git@github.com:|https://github.com/|g' .gitmodules
+
+# Initialize and update submodules (now using HTTPS)
+RUN git submodule update --init --recursive
+
+RUN make -C target/snitch_cluster verilator
+
+RUN make -C target/snitch_cluster DEBUG=ON sw
+
+WORKDIR /
+
+RUN cd /repo/target/snitch_cluster/work-vlt/riscv-isa-sim \
+  && ./configure --prefix=/opt/snitch-spike \
+  && make install
+
+WORKDIR /src
+
+RUN curl -L -o riscv32-snitch-llvm-ubuntu2204-15.0.0-snitch-0.2.0.tar.gz https://github.com/pulp-platform/llvm-project/releases/download/15.0.0-snitch-0.2.0/riscv32-snitch-llvm-ubuntu2204-15.0.0-snitch-0.2.0.tar.gz
+
+RUN tar -xzf riscv32-snitch-llvm-ubuntu2204-15.0.0-snitch-0.2.0.tar.gz -C /src
+
+# # Combine all build steps into a single RUN command to leverage caching.
+# # If the repo URL or commit hash doesn't change, this layer remains cached.
+# RUN git clone --recursive https://github.com/pulp-platform/snitch_cluster.git . \
+#    && git reset --hard c757830632b7f7cd7d8ed511e5cc1eb662b43b37 \
+#    && git submodule update --init --recursive \
+#    # Build the verilator model and software in the same layer
+#    && cd /src/target/snitch_cluster \
+#    && make verilator \
+#    && make DEBUG=ON sw
```

Comment:
We should indeed combine commands above to reduce Docker layers once we have a working image.